circuit/src/test/kotlin/com/slack/circuit/EventListenerTest.kt

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2022 Slack Technologies, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.slack.circuit
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.State
+import androidx.compose.runtime.mutableStateOf
+import app.cash.molecule.RecompositionClock.Immediate
+import app.cash.molecule.moleculeFlow
+import app.cash.turbine.Turbine
+import app.cash.turbine.test
+import com.google.common.truth.Truth.assertThat
+import java.util.concurrent.TimeUnit
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.test.runTest
+import kotlinx.parcelize.Parcelize
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.Timeout
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.annotation.Config
+
+@Config(minSdk = 32, manifest = Config.NONE)
+@RunWith(RobolectricTestRunner::class)
+class EventListenerTest {
+
+  @JvmField
+  @Rule
+  val timeout: Timeout =
+    Timeout.builder().withTimeout(10, TimeUnit.SECONDS).withLookingForStuckThread(true).build()
+
+  @Test
+  fun basicEventRecording() = runTest {
+    val eventListenerFactory = RecordingEventListener.Factory()
+    val state = mutableStateOf("State")
+    val events = MutableSharedFlow<String>()
+
+    val presenter = StringPresenter(state)
+    val ui = StringUi(events)
+    val circuit =
+      Circuit.Builder()
+        .addPresenterFactory { _, _ -> presenter }
+        .addUiFactory { ScreenView(ui) }
+        .eventListenerFactory(eventListenerFactory)
+        .build()
+
+    moleculeFlow(Immediate) {
+        CircuitProvider(circuit) { CircuitContent(TestScreen) }
+        presenter.present(events)
```

Comment:
`presenter.present(events)` can be frustrating with test tooling, because shared flow subscription is racy — whether the event is received or not depends on when the subscriber hooks up to the flow.

Here's a workaround for this problem similar to what we use in our test harness:

```kotlin
    val eventsChannel = Channel<UiEvent>(50)
    lateinit var events : Flow<UiEvent>
    val job = launch { 
      events = eventsChannel.consumeAsFlow().shareIn(this, SharingStarted.Lazily) 
    }

    ...[do test stuff with molecule]...

    job.cancel()
```

(Here is what we ACTUALLY use in our test harness. It is really questionable to me, but I provide it here anyway in the spirit of candor/maybe the code I wrote above is broken somehow:)

```kotlin
    val eventsChannel = Channel<UiEvent>(50)
    val events  = MutableSharedFlow<UiEvent>()
    val job = launch { eventsChannel.consumeAsFlow().collect(events::emit) }

    ...molecule test stuff...
    
    job.cancel()

```